crates/holochain/src/core/state/metadata/activity.rs

```diff
@@ -0,0 +1,201 @@
+use super::*;
+
+/// Rules to make this monotonic
+/// - Invalid overwrites valid and any invalids later in the chain.
+/// - Later Valid headers overwrite earlier Valid.
+/// - If there are two Valid status at the same seq num then insert an Fork.
+pub(super) fn add_chain_status(
+    prev_status: ChainStatus,
+    incoming_status: ChainStatus,
+) -> Option<ChainStatus> {
+    use ChainStatus::*;
+    match (&prev_status, &incoming_status) {
+        (Valid(p), Valid(c)) => {
+            if p.header_seq == c.header_seq && p.hash != c.hash {
+                // Found a fork so insert a fork
+                Some(Forked(ChainFork{
+                    fork_seq: p.header_seq,
+                    first_header: p.hash.clone(),
+                    second_header: c.hash.clone(),
+                }))
+            } else if p == c || p.header_seq > c.header_seq {
+                // Both are the same no need to overwrite or
+                // Previous is more recent so don't overwrite
+                None
+            } else {
+                // Otherwise overwrite with current
+                Some(incoming_status)
+            }
+        }
+        // # Reasons to not overwrite
+        // ## Invalid / Forked where the previous is earlier in the chain
+        (Invalid(p), Forked(c)) if p.header_seq <= c.fork_seq => None,
+        (Invalid(p), Invalid(c)) if p.header_seq <= c.header_seq => None,
+        (Forked(p), Invalid(c)) if p.fork_seq <= c.header_seq => None,
+        (Forked(p), Forked(c)) if p.fork_seq <= c.fork_seq => None,
+        // ## Previous is Invalid / Forked and current is valid
+        (Invalid(_), Valid(_)) | (Forked(_), Valid(_))
+        // Current is empty
+        | (_, Empty) => None,
+        // Previous should never be empty
+        (Empty, _) => unreachable!("Should never cache an empty status"),
+        // The rest are reasons to overwrite
+        _ => Some(incoming_status),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use fixt::prelude::*;
+    use holo_hash::fixt::*;
+    use ChainStatus::*;
+
+    #[test]
+    fn add_chain_status_test() {
```

Comment:
My comment about having test stubs on the other PR wasn't so much about actually writing the test, but about starting to accumulate spec-like docs that can be parsed into a single document, which I would still like. I suggested the `/// TEST: blah blah` convention in the other PR. Maybe we can have a way to aggregate comments scattered throughout the test too, though, like you have here? That might be even more readable.

I kind of like this convention here of `// - text` meaning that the text will be added as a bullet point. But then there should probably be some descriptive text at the top, and maybe a marker which declares that the `// - ` items will be parsed. Maybe:
 
```suggestion
    /// TEST: The following chain status transitions have the proper precedence
    #[test]
    fn add_chain_status_test() {
```

Which would parse to something like this, in the yet-to-be-written parser:

```
add_chain_status_test
The following chain status transitions have the proper precedence:
- Invalid overwrites valid
- Invalid overwrites any invalids later in the chain.
- Invalid overwrites any forks later in the chain
```

etc.

Also, sorry to be targeting your PR for this unrelated thing, but testing is not related to any PR in particular, and we have to start somewhere. If we think this is a good idea, that is.

Reply:
Here's some thoughts on this idea: https://hackmd.io/wnj3HdRHQzK5584KosOkkg

Reply:
I like this idea. I still think we should try and factor out and actually write unit tests where we can